src/document/json/tree.ts

```diff
@@ -198,7 +249,41 @@ export class Tree {
       throw new Error('it is not initialized yet');
     }
 
-    this.tree.onChanges(handler);
+    this.tree.onChangeCollect((changes: Array<TreeChange>) => {
+      (this.changes as Array<TreeChange>).push(...changes);
+    });
+    this.tree.onChanges(() => {
+      handler(this.changes as Array<TreeChange>);
+
+      this.changes = [];
+    });
+  }
+
+  /**
+   * `onChangesByPath` registers a handler of onChanges event.
+   */
+  onChangesByPath(handler: (changes: Array<TreeChangeWithPath>) => void): void {
+    if (!this.context || !this.tree) {
+      throw new Error('it is not initialized yet');
+    }
+    const collect = (changes: Array<TreeChange>) => {
+      const changeWithPath = changes.map(({ from, to, ...rest }) => {
+        return {
+          from: this.tree!.indexToPath(from),
+          to: this.tree!.indexToPath(to),
+          ...rest,
+        };
+      });
+
+      (this.changes as Array<TreeChangeWithPath>).push(...changeWithPath);
+    };
+
+    this.tree.onChangeCollect(collect);
+    this.tree.onChanges(() => {
+      handler(this.changes as Array<TreeChangeWithPath>);
+
+      this.changes = [];
+    });
```

Comment:
How about narrowing down the scope of changes buffer in a local variable?

```suggestion
  onChangesByPath(handler: (changes: Array<TreeChangeWithPath>) => void): void {
    if (!this.context || !this.tree) {
      throw new Error('it is not initialized yet');
    }
    const localChanges: Array<TreeChangeWithPath> = [];
    const collect = (changes: Array<TreeChange>) => {
      const changeWithPath = changes.map(({ from, to, ...rest }) => {
        return {
          from: this.tree!.indexToPath(from),
          to: this.tree!.indexToPath(to),
          ...rest,
        };
      });

      localChanges.push(...changeWithPath);
    };

    this.tree.onChangeCollect(collect);
    this.tree.onChanges(() => {
      handler(localChanges);
      localChanges.length = 0;
    });
  }
```